packages/itwin/tree-widget/src/tree-widget-react/components/trees/classifications-tree/ClassificationsTreeDefinition.ts

```diff
@@ -330,3 +354,340 @@ function createClassificationHasChildrenSelector(classificationAlias: string) {
+          classificationIds: new Array<ClassificationId>(),
+          element2dIds: new Array<ElementId>(),
+          element3dIds: new Array<ElementId>(),
+        },
+      ),
+      mergeMap((ids) => {
+        const elements2dLength = ids.element2dIds.length;
+        const elements3dLength = ids.element3dIds.length;
+        return merge(
+          from(ids.classificationTableIds).pipe(
+            map((id): HierarchyFilteringPath => ({ path: [{ id, className: CLASS_NAME_ClassificationTable }], options: { autoExpand: true } })),
+          ),
+          idsCache.getClassificationsPathObs(ids.classificationIds).pipe(map((path) => ({ path, options: { autoExpand: true } }))),
+          from(ids.element2dIds).pipe(
+            bufferCount(Math.ceil(elements2dLength / Math.ceil(elements2dLength / 5000))),
+            releaseMainThreadOnItemsCount(1),
+            mergeMap((block) => createGeometricElementInstanceKeyPaths({ idsCache, imodelAccess, targetItems: block, type: "2d" }), 10),
+          ),
+          from(ids.element3dIds).pipe(
+            bufferCount(Math.ceil(elements3dLength / Math.ceil(elements3dLength / 5000))),
+            releaseMainThreadOnItemsCount(1),
+            mergeMap((block) => createGeometricElementInstanceKeyPaths({ idsCache, imodelAccess, targetItems: block, type: "3d" }), 10),
+          ),
+        );
+      }),
+    );
+  };
+}
+
+function createGeometricElementInstanceKeyPaths(props: {
+  idsCache: ClassificationsTreeIdsCache;
+  imodelAccess: ECClassHierarchyInspector & LimitingECSqlQueryExecutor;
+  targetItems: Id64Array;
+  type: "2d" | "3d";
+}): Observable<HierarchyFilteringPath> {
+  const { targetItems, imodelAccess, type, idsCache } = props;
+  if (targetItems.length === 0) {
+    return EMPTY;
+  }
+
+  const separator = ";";
+
+  return defer(() => {
+    const ctes = [
+      `ElementsHierarchy(ECInstanceId, ParentId, Path) AS (
+        SELECT
+          e.ECInstanceId,
+          e.Parent.Id,
+          'e${type}${separator}' || CAST(IdToHex([e].[ECInstanceId]) AS TEXT)
+        FROM  ${type === "2d" ? CLASS_NAME_GeometricElement2d : CLASS_NAME_GeometricElement3d} e
+        WHERE e.ECInstanceId IN (${targetItems.join(",")})
+
+        UNION ALL
+
+        SELECT
+          pe.ECInstanceId,
+          pe.Parent.Id,
+          'e${type}${separator}' || CAST(IdToHex([pe].[ECInstanceId]) AS TEXT) || '${separator}' || ce.Path
+        FROM ElementsHierarchy ce
+        JOIN ${type === "2d" ? CLASS_NAME_GeometricElement2d : CLASS_NAME_GeometricElement3d} pe ON pe.ECInstanceId = ce.ParentId
+      )`,
+    ];
+    const ecsql = `
+      SELECT
+        IIF(c.Parent.Id IS NULL,
+          'ct${separator}' || CAST(IdToHex(c.Model.Id) AS TEXT) || '${separator}c${separator}' || CAST(IdToHex([c].[ECInstanceId]) AS TEXT) || '${separator}' || e.Path,
+          'c${separator}' || CAST(IdToHex([c].[ECInstanceId]) AS TEXT) || '${separator}' || e.Path
+        ),
```

Comment:
why are we combining this into a string? wouldn't it be more convenient to select the values as separate SELECT columns?